File: a524(phone_puzzle)/Python/phone_puzzle.py
from sys import stdin, stdout
from typing import List


num_of_digit: int or str = 0
password: List[int] = [0, 0, 0, 0, 0, 0, 0, 0]
# won't use index 0
unuse_digit: List[bool] = [True, True, True, True, True, True, True, True, True]


def try_next_password(prev_try: int, pos: int):
    password[pos] = prev_try
    unuse_digit[prev_try] = False

    if pos == num_of_digit - 1:
        for word in password[:num_of_digit]:
            stdout.write(str(word))
        stdout.write('\n')
        return

    for i in range(num_of_digit, 0, -1):
        if unuse_digit[i]:
            try_next_password(i, pos + 1)
            unuse_digit[i] = True
    return


while True:
    num_of_digit = stdin.readline().strip('\n')
    if not num_of_digit:
        break

    num_of_digit = int(num_of_digit)
    for i in range(num_of_digit, 0, -1):
        try_next_password(i, 0)
        # cancel the mark
        unuse_digit[i] = True

File: a524(phone_puzzle)/C++/phone_puzzle.cpp
#include <iostream>
#include <vector>


void TryNextPassword(const int num_of_digit,
                     std::vector<bool>& unuse_digit, std::vector<int>& password,
                     const int prev_try, const int pos) {
  password.at(pos) = prev_try;
  unuse_digit.at(prev_try) = false;
  if (pos == num_of_digit - 1) {
    for (int word : password) {
      std::cout << word;
    }
    std::cout << '\n';
    return;
  }
  for (int i = num_of_digit; i > 0; --i) {
    if (unuse_digit.at(i)) {
      TryNextPassword(num_of_digit, unuse_digit, password, i, pos + 1);
      unuse_digit.at(i) = true;
    }
  }
  return;
}


int main(int argc, char const *argv[]) {
  std::ios_base::sync_with_stdio(false);
  std::cin.tie(0);
  
  int num_of_digit = 0;
  std::vector<int> password;
  while (std::cin >> num_of_digit) {
    password.resize(num_of_digit);

    // won't use at(0)
    std::vector<bool> unuse_digit(num_of_digit + 1, true);
    for (int i = num_of_digit; i > 0; --i) {
      TryNextPassword(num_of_digit, unuse_digit, password, i, 0);
      unuse_digit.at(i) = true;
    }
  }

  return 0;
}
